### PR TITLE
Smartmontools prefab

### PR DESCRIPTION
**Pull request by chrisvdg**
Manual install prefab for smartctl 6.6

Needed for:
https://github.com/Jumpscale/lib9/pull/222

### CODE REVIEW

File: modules/monitoring/PrefabSmartmontools.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+class PrefabSmartmontools(base):
+    """
+    Prefab module for smartmontools (smartctl command)
+    """
+
+    NAME = "smartmontools"
+
+    _DOWNLOAD_URL = "https://sourceforge.net/projects/smartmontools/files/smartmontools/6.6/smartmontools-6.6.tar.gz/download"
+    _DOWNLOAD_LOCATION = "/tmp/smartlibtools.tar.gz"
```
Comment (zaibon) [inline modules/monitoring/PrefabSmartmontools.py]: don't hardcode path like this, build it using `j.dirs.TMPDIR`

File: modules/monitoring/PrefabSmartmontools.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        # cleanup
+        cmd = ("rm -rf %s; rm -rf %s" % (self._DOWNLOAD_LOCATION, self._EXTRACT_LOCATION))
+        rc, out, err = self.prefab.executor.execute(cmd, die=False)
+        if rc != 0:
+            self._node.logger.error("Failed to cleanup smartmontools installation: %s", err)
+
+        if not self._check_smartctl():
+            raise RuntimeError("Failed to install Smartmontools")
+    
+    def _check_smartctl(self):
```
Comment (zaibon) [inline modules/monitoring/PrefabSmartmontools.py]: This methods only check if smartctl is installed or not, this check is part of the base class of each prefab module: https://github.com/Jumpscale/prefab9/blob/master/JumpScale9Prefab/PrefabBase.py#L176

Also the problem is you just check if smartctl is installed or not, but not that is it has the right version.
So line 26, you could just get true, while the version installed is not the one you want.

File: modules/monitoring/PrefabSmartmontools.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        Version 6.6 is needed for NVMe support)
+        Ensures smartctl is installed
+        """
+        if self._check_smartctl():
+            self.logger.info("smartctl is present")
+            return
+            
+        self.logger.info("installing smartctl...")
+
+        cmd = "wget %s -O %s" % (self._DOWNLOAD_URL, self._DOWNLOAD_LOCATION)
```
Comment (zaibon) [inline modules/monitoring/PrefabSmartmontools.py]: maybe you look for : `j.tools.prefab.local.core.file_download`

File: modules/monitoring/PrefabSmartmontools.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        rc, out, err = self.prefab.executor.execute(cmd, die=False)
+        if rc != 0:
+            raise RuntimeError("Error downloading smartmontools:\n%s\n\n%s" % (
+                out, err))
+
+        cmd = "mkdir -p %s && tar -xf %s -C %s --strip-components 1" % (
```
Comment (zaibon) [inline modules/monitoring/PrefabSmartmontools.py]: extraction of tgz is also supported by `j.tools.prefab.local.core.file_download`

File: modules/monitoring/PrefabSmartmontools.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            self._EXTRACT_LOCATION, self._DOWNLOAD_LOCATION, self._EXTRACT_LOCATION)
+        rc, out, err = self.prefab.executor.execute(cmd, die=False)
+        if rc != 0:
+            raise RuntimeError("Error extracting smartmontools:\n%s\n\n%s" % (
+                out, err))
+
+        cmd = "cd %s && ./configure && make && make install" % (self._EXTRACT_LOCATION)
+        rc, out, err = self.prefab.executor.execute(cmd, die=False)
```
Comment (zaibon) [inline modules/monitoring/PrefabSmartmontools.py]: no need to use the executor directly prefab has its own method to run a command: `self.prefab.core.run`

File: modules/monitoring/PrefabSmartmontools.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    """
+
+    NAME = "smartmontools"
+
+    _DOWNLOAD_URL = "https://sourceforge.net/projects/smartmontools/files/smartmontools/6.6/smartmontools-6.6.tar.gz/download"
+    _DOWNLOAD_LOCATION = "/tmp/smartlibtools.tar.gz"
+    _EXTRACT_LOCATION = "/tmp/smartlibtools"
+
+
+    def install_smartctl(self):
```
Comment (zaibon) [inline modules/monitoring/PrefabSmartmontools.py]: This method is called installed, while we are actually building. Maybe would be better to call it build.
Also no need to suffix the method name with smartctl, you are already in the smartctl prefab module, so the we know what we are dealing with.

Comment (zaibon) [reply modules/monitoring/PrefabSmartmontools.py]: This make me think, doesn't smartctl provide a pre-build version ?
Do we really have to compile this from source ?

Comment (chrisvdg) [reply modules/monitoring/PrefabSmartmontools.py]: Don't seem to provide for 16.04 not v6.6 for Ubuntu
https://www.smartmontools.org/wiki/Packages

Comment (chrisvdg) [reply modules/monitoring/PrefabSmartmontools.py]: https://www.smartmontools.org/wiki/Download

**Pull request merged**